src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -751,11 +786,18 @@ internal void ConfigureForJsonSerializer()
                 ThrowHelper.ThrowInvalidOperationException_JsonSerializerIsReflectionDisabled();
             }
 
-            MakeReadOnly();
+            Debug.Assert(_typeInfoResolver != null);
```

Comment:
This method is reading _typeInfoResolver multiple times. What happens if another thread sets it in the interim? I'm on my phone so haven't reviewed all the code, but how is this all thread-safe?

Reply:
It generally speaking isn't, for example if you try to run `JsonSerializer.Serialize(x, options);` and `options.TypeInfoResolver = null;` calls concurrently that particular assertion will fail. We've been regarding that particular condition as acceptable, if you try to both configure an options instance and serialize with it concurrently you should expect undefined behavior. 

What should be thread-safe though is concurrent calls to `JsonSerializer.Serialize(x, options);` which is I wasn't able to achieve in the `System.Net.Http.Json` layer.